src/clean-core/capped_vector.hh

```diff
@@ -237,7 +230,40 @@ public:
     }
 
 private:
-    using compact_size_t = detail::compact_size_t_typed<T, N>;
+    static void _move_range(T* src, compact_size_t num, T* dest)
+    {
+        if constexpr (std::is_trivially_move_constructible_v<T> && std::is_trivially_copyable_v<T>)
+        {
+            if (num > 0)
+                std::memcpy(dest, src, sizeof(T) * num);
+        }
+        else
+        {
+            for (compact_size_t i = 0; i < num; ++i)
+                new (placement_new, &dest[i]) T(cc::move(src[i]));
+        }
+    }
+    static void _copy_range(T const* src, compact_size_t num, T* dest)
+    {
+        if constexpr (std::is_trivially_copyable_v<T>)
+        {
+            if (num > 0)
+                std::memcpy(dest, src, sizeof(T) * num);
+        }
+        else
+        {
+            for (compact_size_t i = 0; i < num; ++i)
+                new (placement_new, &dest[i]) T(src[i]);
+        }
+    }
+    static void _destroy_reverse(T* data, compact_size_t size, compact_size_t to_index = 0)
+    {
+        if constexpr (!std::is_trivially_destructible_v<T>)
+        {
+            for (compact_size_t i = size; i > to_index; --i)
+                data[i - 1].~T();
+        }
+    }
```

Comment:
We used those in a few places now I think it's time for a `detail/container_impl_helper.hh` or something ;)